### ISSUE TITLE
Version 11.6.0 not working session storage (invalid_site error)

### ISSUE THREAD

**Issue by overallduka** [Shopify/shopify_app#860]
Hey!

So, I'm having a big issue here, I am trying to update my app to 11.6.0 to have the `SameSite=None; Secure`, but I have a session error exactly on this version, I have sure about this because I tried with 11.0.0, 11.4.0 and 11.5.1 and the error just happens on 11.6.0.

The error is this:

> I, [2020-01-16T20:08:24.130599 #38014]  INFO -- omniauth: (shopify) Request phase initiated.
> E, [2020-01-16T20:08:24.131116 #38014] ERROR -- omniauth: (shopify) Authentication failure! invalid_site encountered.
> 
> OmniAuth::Error (invalid_site):
> 
> omniauth (1.9.0) lib/omniauth/failure_endpoint.rb:25:in `raise_out!'
> omniauth (1.9.0) lib/omniauth/failure_endpoint.rb:20:in `call'
> omniauth (1.9.0) lib/omniauth/failure_endpoint.rb:12:in `call'
> omniauth (1.9.0) lib/omniauth/strategy.rb:491:in `fail!'
> omniauth-shopify-oauth2 (2.2.0) lib/omniauth/strategies/shopify.rb:113:in `request_phase'
> omniauth (1.9.0) lib/omniauth/strategy.rb:226:in `request_call'
> omniauth (1.9.0) lib/omniauth/strategy.rb:188:in `call!'
> omniauth (1.9.0) lib/omniauth/strategy.rb:169:in `call'
> newrelic_rpm (6.7.0.359) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
> omniauth (1.9.0) lib/omniauth/builder.rb:64:in `call'
> newrelic_rpm (6.7.0.359) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
> newrelic_rpm (6.7.0.359) lib/new_relic/rack/agent_hooks.rb:30:in `traced_call'
> newrelic_rpm (6.7.0.359) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
> newrelic_rpm (6.7.0.359) lib/new_relic/rack/browser_monitoring.rb:32:in `traced_call'
> newrelic_rpm (6.7.0.359) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
> rack (2.0.7) lib/rack/tempfile_reaper.rb:15:in `call'
> newrelic_rpm (6.7.0.359) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
> rack (2.0.7) lib/rack/etag.rb:25:in `call'
> newrelic_rpm (6.7.0.359) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
> rack (2.0.7) lib/rack/conditional_get.rb:25:in `call'

I did some investigation.. 

For all this versions(11.0.0, 11.4.0, 11.5.1) I have debug the file `config/initializers/omniauth.rb`

This line: `shopify_auth_params = strategy.session['shopify.omniauth_params']&.with_indifferent_access`

Basically for all versions 11.0.0, 11.4.0, 11.5.1 the value of `shopify_auth_params` is getting the shop parameter, the value is:

`{"shop"=>"myshop.myshopify.com"}`

But, for 11.6.0, `shopify_auth_params ` is just `nil`, and that seems to be the cause of the error, why there is the error I don't have idea =/.

Just for the sake of information, I'm using `ShopifyApp::SessionStorage` on the `Shop` model as session storage. But I have tried using `config.session_repository = 'ShopifyApp::InMemorySessionStore'` but don't work either.

Thank you guys for the awesome job on this piece of code and I really hope we can solve this until Chrome make it mandatory to use version 11.6.0(SameSite=None; Secure).

**Comment by tanema** [comment]
I had this issue as well my self for a little while, while testing out my own app. Try updating the gem again. Run bundler and restart your app and it might work. I never was able to narrow down what this issue was.

**Comment by tanema** [comment]
Also just try clearing your cookies etc, like I said I had this problem while testing and it didnt lead to anything and it suddenly started working while I was trying to analyze the issue. If you are able to get it working (or if you are not) let me know the steps you took

**Comment by overallduka** [comment]
Hey, sorry about the delay, yet trying to implement the new version 11.6.0. 

The issue with the domain being blank seems to happen just on localhost, in localhost basically won't work, but when live it works.

The error on localhost is the same:

![6E2B7xOw](https://user-images.githubusercontent.com/4923856/73297682-c450be00-41ea-11ea-8344-de677709f9ee.png)

**Comment by overallduka** [comment]
Actually, I could solve the issue, after check this https://github.com/Shopify/shopify_app/issues/874 I have tried this:

```
ShopifyApp.configure do |config|
...
  config.enable_same_site_none = Rails.env.production?
end
```

And it works locally.

**Comment by tanema** [comment]
No that just disabled the middleware and that is not a good fix because then your app will break on Feb 4th.

An internal developer had this problem and solved it by running bundle install again. It looks like it updated the omniauth dependencies and it worked

**Comment by tjoyal** [comment]
Each time I spawn up an app I haven't use in a while I end up doing the same mistakes (including today). Might as well share a bit. I'm not shopify-app expert, so your mileage may vary.

This error I believe is due to browsers preventing setting cookies on a redirect that target localhost? That's more of a guess, it wouldn't be crazy as I know Safari has some strange rules about domain transfer and cookies.

Even if you power through that with the `enable_same_site_none` flag, you'll still experience app issues down the road using localhost as you won't be able to load your app within the iframe of Shopify admin.

I suppose you could set an entry in your local dns to trick the browser to think your local environment is not local. I spawned up a ngrok tunnel and both problems went away.

**Comment by tanema** [comment]
It is no longer possible to use localhost to develop a shopify app. Since SameSite=none is required for chrome, the secure attribute is also required on each cookie. To have the secure attribute valid on each cookie, you need to be serving it from https/ssl.

**Comment by sshaw** [comment]
@tjoyal to develop on localhost I setup a dev version of my app in Shopify that's not embedded.

**Comment by aarkerio** [comment]
I solved the issue adding a middleware, search for "SameSiteCookieMiddleware rails"

**Comment by tanema** [comment]
shopify_app has samesite cookie middle ware built in by default.

**Comment by shopify01** [comment]
I am facing this issue in my production environment. I have an http as well as https url set as whitelisted urls in the app but the requests, for some reason, are always redirected to http url causing invalid_site error.

**Comment by diei** [comment]
I have the some problem with shopify_app 14.4.3. I use ngrok to route to my local webservice. When I try to link a development shop to my custom (local) app, I get always the `invalid_site` error.

I just used the shopify_app generators and added `enable_same_site_none = false` to config.

My Shopify config is:
```
ShopifyApp.configure do |config|
  config.application_name = "My App"

  config.api_key = ENV['SHOPIFY_API_KEY']
  config.secret = ENV['SHOPIFY_API_SECRET']

  config.old_secret = ""
  config.scope = "write_assigned_fulfillment_orders,merchant_managed_fulfillment_orders,third_party_fulfillment_orders,read_products"
  config.embedded_app = false
  config.after_authenticate_job = false
  config.api_version = "2020-07"
  config.shop_session_repository = 'Shop'
  config.allow_jwt_authentication = true
  config.enable_same_site_none = false
end
```

**Comment by tanema** [comment]
Can you share your omniauth initializer and/or what you are doing to get this error (steps to reproduce)? Invalid site happens both on the request phase and callback phase if the host that you are requesting authentication from and receiving the callback from is not a shopify domain (ending in 'myshopify.com')
